docs/spec/tendermint-fork-cases/TendermintAcc3.tla

```diff
@@ -0,0 +1,447 @@
+ of the Tendermint paper: https://arxiv.org/abs/1807.04938
+
+ For the purposes of fork accountability, the faulty processes are partitioned
+ into two sets: the Byzantine processes and the defective processes.
+ While the Byzantine processes can demonstrate arbitrary behavior, including
+ no communication, the defective processes send their messages but deviate
+ from the protocol in two ways:
+
+   - Equivocation: a defective process may send two different values
+     in the same round.
+
+   - Amnesia: a defective process may lock a value, although it has locked
+     another value in the past.
+
+ TODO: as there is no clear difference between Byzantine and Defective processes anymore,
+ we will probably merge them together.
+
+ * Version 3. Modular and parameterized definitions.
+ * Version 2. Bugfixes in the spec and an inductive invariant.
+ * Version 1. A preliminary specification.
+
+ Zarko Milosevic, Igor Konnov, 2019-2020.
+ *)
+
+EXTENDS Integers, FiniteSets
+
+(********************* PROTOCOL PARAMETERS **********************************)
+CONSTANTS
+    Corr,          \* the set of correct processes 
+    Defective,     \* the set of processes that show defects, e.g, amnesia or equivocation
+    Byzantine,     \* the set of Byzantine processes, may be empty
+    N,             \* the total number of processes: correct, defective, and Byzantine
+    T,             \* an upper bound on the number of Byzantine processes
+    ValidValues,   \* the set of valid values, proposed both by correct and faulty
+    InvalidValues, \* the set of invalid values, never proposed by the correct ones
+    MaxRound,      \* the maximal round number
+    Proposer       \* the proposer function from 0..NRounds to 1..N
+
+ASSUME(N = Cardinality(Corr \union Defective \union Byzantine))
+
+(*************************** DEFINITIONS ************************************)
+Faulty == Defective \union Byzantine  \* the set of faulty processes
+AllProcs == Corr \union Faulty      \* the set of all processes
+Rounds == 0..MaxRound               \* the set of potential rounds
+NilRound == -1   \* a special value to denote a nil round, outside of Rounds
+RoundsOrNil == Rounds \union {NilRound}
+Values == ValidValues \union InvalidValues \* the set of all values
+NilValue == "None"  \* a special value for a nil round, outside of Values
+ValuesOrNil == Values \union {NilValue}
+
+\* a value hash is modeled as identity
+Id(v) == v
+
+\* The validity predicate
+IsValid(v) == v \in ValidValues
+
+\* the two thresholds that are used in the algorithm
+THRESHOLD1 == T + 1     \* at least one process is not faulty
+THRESHOLD2 == 2 * T + 1 \* a quorum when having N > 3 * T
+
+(********************* TYPE ANNOTATIONS FOR APALACHE **************************)
+\* the operator for type annotations
+a <: b == a
+
+\* the type of message records
+MT == [type |-> STRING, src |-> STRING, round |-> Int,
+       proposal |-> STRING, validRound |-> Int, id |-> STRING]
+       
```

Comment:
Some comments on what the message record fields encode would be useful. For example, it is not clear that `src` is a process ID or that proposal is one of `"PROPOSAL", "PREVOTE", "PRECOMMIT"`. Not sure if this is the appropriate place, maybe they can go in the file where the invariants are defined.

Reply:
Right. This will change with the new type checker anyways.